mountpoint-s3-fs/src/manifest/builder.rs

```diff
@@ -1,62 +1,392 @@
-use super::ManifestError;
+use std::collections::HashSet;
+use std::io::BufReader;
+use std::path::{Path, PathBuf};
+use std::{collections::HashMap, fs::File};
+
+use fuser::FUSE_ROOT_ID;
+use serde::Deserialize;
+
+use crate::fs::InodeKind;
+use crate::metablock::{ValidKey, ValidKeyError, ValidName};
+use crate::prefix::Prefix;
+use crate::s3::config::S3Path;
 use crate::{
-    manifest::CsvReader,
     manifest::db::{Db, DbEntry},
-    metablock::ValidName,
+    manifest::{CsvReader, ManifestError},
 };
-use std::fs::File;
-use std::io::BufReader;
-use std::path::Path;
 
-pub fn create_db(
+#[derive(Debug, Deserialize)]
+pub struct ChannelConfig {
+    pub directory_name: String,
+    pub bucket_name: String,
+    #[serde(default)]
+    pub prefix: String,
+    pub manifest_path: PathBuf,
+}
+
+pub struct ChannelManifest<I> {
+    pub directory_name: String,
+    pub s3_path: S3Path,
+    pub entries: I,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InputManifestEntry {
+    partial_key: ValidKey, // guaranteed to be InodeKind::File (=> not empty)
+    etag: String,
+    size: usize,
+}
+
+impl InputManifestEntry {
+    pub fn new(partial_key: &str, etag: &str, size: usize) -> Result<Self, ManifestError> {
+        Self::from_owned(partial_key.to_string(), etag.to_string(), size)
+    }
+
+    pub fn from_owned(partial_key: String, etag: String, size: usize) -> Result<Self, ManifestError> {
+        let partial_key = ValidKey::try_from(partial_key)?;
+        if partial_key.is_empty() {
+            return Err(ManifestError::InvalidKey(ValidKeyError::InvalidKey(
+                partial_key.to_string(),
+            )));
+        }
+        if partial_key.kind() != InodeKind::File {
+            return Err(ManifestError::FolderMarker(partial_key.to_string()));
+        }
+        Ok(Self {
+            partial_key,
+            etag,
+            size,
+        })
+    }
+
+    fn into_db_entry(
+        self,
+        id: u64,
+        parent_id: u64,
+        channel_id: usize,
+        parent_partial_key: ValidKey,
+    ) -> Result<DbEntry, ManifestError> {
+        debug_assert_eq!(self.partial_key.kind(), InodeKind::File);
+
+        // parse and validate the name
+        let name = match self.partial_key.rsplit_once('/') {
+            Some((_, name)) => name.to_string(),
+            None => self.partial_key.to_string(),
+        };
+        let _ = ValidName::parse_str(&name)
+            .map_err(|_| ManifestError::InvalidKey(ValidKeyError::InvalidKey(name.clone())))?;
```

Comment:
Can't we just get the name?
```suggestion
        let name = self.partial_key.name().to_string();
```